ultravox/data/datasets.py

```diff
@@ -1060,43 +1068,73 @@ def create_dataset(name: str, args: VoiceDatasetArgs) -> data.IterableDataset:
 
 
 class InterleaveDataset(data.IterableDataset):
-    """Interleaves multiple IterableDataset objects."""
+    """Interleaves multiple IterableDataset objects based on normalized weights."""
 
     def __init__(
-        self, datasets: Sequence[data.IterableDataset], repeat: bool = False
+        self,
+        datasets: Sequence[data.IterableDataset],
+        stop_strategy: str = "last_exhausted",
+        seed: Optional[int] = 42,
+        static: bool = False,
     ) -> None:
         """
         Args:
-            datasets: a list of IterableDataset objects
-            repeat: whether to repeat the datasets indefinitely.
-                This matters most when the datasets have different lengths.
-                Let's say you have two datasets, A and B which have 5 and 3 samples respectively.
-
-                `repeat=False`: [A0, B0, A1, B1, A2, B2, A3, A4]
-                `repeat=True` : [A0, B0, A1, B1, A2, B2, A3, B0, A4, B1, A0, ...]
-
-                NOTE: with `repeat=True`, `__iter__` never stops.
+            datasets: A list of data.IterableDataset objects.
+            stop_strategy: Strategy for stopping iteration. Choose from 'first_exhausted', 'last_exhausted', 'never_stop'.
+            seed: Optional seed for reproducibility.
+            static: If true, the datasets are interleaved in a static order with equal weights.
         """
         super().__init__()
         self._datasets = datasets
-        self._repeat = repeat
+        self._rng = np.random.default_rng(seed)
+        self._static = static
+
+        if stop_strategy not in ["first_exhausted", "last_exhausted", "never_stop"]:
+            raise ValueError(
+                "Invalid stop_strategy. Choose from 'first_exhausted', 'last_exhausted', 'never_stop'."
+            )
+        self._stop_strategy = stop_strategy
+
+        weights = [getattr(ds, "get_weight", lambda: 1)() for ds in datasets]
```

Comment:
Multiple things don't seem to match up on this line. Are you sure it works correctly?
```suggestion
        weights = [getattr(ds, "weight", 1) for ds in datasets]
```